tuareg.el

```diff
@@ -3645,35 +3682,43 @@ It is assumed that the range START-END delimit valid OCaml phrases."
 (defun tuareg-eval-region (start end)
   "Eval the current region in the OCaml REPL."
   (interactive "r")
-  (setq tuareg-interactive-last-phrase-pos-in-source start)
-  (save-excursion
-    (goto-char start)
-    (tuareg-backward-beginning-of-defun)
-    (setq start (point))
-    (setq end (cdr (tuareg-region-of-defun end)))
-    (if end
-        (tuareg-interactive--send-region start end)
-      (message "The expression after the point is not well braced."))))
+  (tuareg-interactive--send-region start end))
 
 (define-obsolete-function-alias 'tuareg-narrow-to-phrase #'narrow-to-defun
   "Apr 10, 2019")
 
 (defun tuareg-eval-phrase ()
-  "Eval the surrounding OCaml phrase (or block) in the OCaml REPL."
+  "Eval the surrounding OCaml phrase (or block) in the OCaml REPL.
+If the region is active, evaluate all phrases intersecting the region."
   (interactive)
-  (let ((opoint (point)))
-    ;; Move before the comment, if we are in one.
-    (let ((ppss (syntax-ppss)))
-      (if (nth 4 ppss) (goto-char (- (nth 8 ppss) 1))))
-    (let ((phrase (tuareg-discover-phrase)))
-      (if phrase
-          (progn
-            (tuareg-interactive--send-region (car phrase) (cadr phrase))
-            (if tuareg-skip-after-eval-phrase
-              (progn (goto-char (car (cddr phrase)))
-                     (tuareg-skip-blank-and-comments))
-              (goto-char opoint)))
-        (message "The expression after the point is not well braced.")))))
+  (let ((opoint (point))
+        start end)
+    (cond
+     ((region-active-p)
+      (setq start (min (point) (mark)))
+      (setq end (max (point) (mark)))
+      ;; Extend the region at the endpoints if they are in a phrase.
+      (dolist (pos (list start end))
+        (save-excursion
+          (goto-char pos)
+          (let ((phrase (tuareg-discover-phrase)))
+            (when (and phrase (> (cadr phrase) pos))
+              (setq start (min start (car phrase)))
+              (setq end (max end (cadr phrase))))))))
+     (t
+      ;; Move before the comment, if we are in one.
+      (let ((ppss (syntax-ppss)))
+        (if (nth 4 ppss) (goto-char (- (nth 8 ppss) 1))))
```

Comment:
@monnier (by email): “Actually, I was wondering: why decrement by 1?  I think we should either not decrement by 1 or add a comment explaining why we decrement.”

Reply:
If you like, but the resulting byte-code will be identical so we might just as well write what's more natural. I have no opinion on what that would be in a case like this and will change it to `1-`.

As examples as to when one or the other would be preferable as matter of style, I would write:
```
(setq i (1- i))
```
since that focus on the single-step decrement, but
```
(fun (- x 1) (- x 2))
```
for homogeneity... well, you get the idea.

Reply:
I'll leave the question whether to decrement at all for now since I wasn't privy to the decision. Elucidate me!

Reply:
Actually you are right, that subtraction seems to be completely pointless and I'm removing it.

Reply:
Actually, I think it is no longer necessary: `tuareg-discover-phrase` calls [`tuareg-region-of-defun`](https://github.com/ocaml/tuareg/blob/master/tuareg.el#L2776) which in turn calls [`tuareg-backward-beginning-of-defun`](https://github.com/ocaml/tuareg/blob/master/tuareg.el#L2544) which already [performs the above](https://github.com/ocaml/tuareg/blob/master/tuareg.el#L2556).

Reply:
I'm just wondering for the following code:
```ocaml
(* Comment 1 *)
let x =
  (* Comment 2 *)
  2

(* Comment 3 *)
let y = 2
```
what should the phrase be if the cursor is in `Comment 3` (at the moment it is the `x` one)?

Reply:
Yes, but that doesn't have anything to do with comments really - if the cursor is on a blank line immediately before a definition, it will still eval the previous one. It's how the phrase discovery works, and we may want to tweak the heuristics for maximum convenience but I'd like to move that outside this PR.

Reply:
> It's how the phrase discovery works, and we may want to tweak the heuristics for maximum convenience but I'd like to move that outside this PR.

I agree.  I was just thinking that it may surprise the user as he thinks the comment is part of the definition of `y` (which is not the case with a blank line).